### PR TITLE
Zero copy support clustered file system(CFS)

### PR DESCRIPTION
**Pull request by marising**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

1. Support NFS storage
2. Read buffer and write buffer for NFS
3. Every replica of the same shard share the same data, that is zero copy
4. Use the same local cache and metadata logic as s3 in the framework

### Documentation entry for user-facing changes

- Motivation: Our company has a distributed file system like AWS EFS, NFS has better performance than S3 protocol after testing.

- Example use: 
1. Mount the same EFS(or other DFS) path to the local machine of every replica in the same shard
2. Define disk and storage policy in config.xml
6. Create table with NFS storage policy
7. Use this table just like the local storage policy

- Reference
https://github.com/ClickHouse/ClickHouse/issues/31113
https://github.com/ClickHouse/ClickHouse/discussions/33998
https://github.com/ClickHouse/ClickHouse/issues/35708

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 8588fe1d9c206d9e766c89b378d1cf9fc81d1fdf with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53629/8588fe1d9c206d9e766c89b378d1cf9fc81d1fdf/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🔴 failure</td></tr>
</table>

**Comment by den-crane** [comment]
><type>nfs</type>

May be better to use `CFS` acronym? https://en.wikipedia.org/wiki/Clustered_file_system

**Comment by marising** [comment]
> > nfs
> 
> Maybe better to use `CFS` acronym? https://en.wikipedia.org/wiki/Clustered_file_system

Thanks for your suggestion, CFS is better.

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 1]
```diff
@@ -1294,6 +1294,54 @@ In this sample configuration:
 </clickhouse>
 ```
 
+## CFS storage {#cfs-storage}
+
+First mount the remote CFS path to the local host such as `/mnt/cfs`.
```
Comment (den-crane) [inline docs/en/engines/table-engines/mergetree-family/mergetree.md:1299]: I would add at least one sentence about what is CFS (even link to wikipedia) and mention a couple: NFS, CephFS, Lustre

Comment (marising) [reply docs/en/engines/table-engines/mergetree-family/mergetree.md:1305]: I added a few simple sentences

**Comment by marising** [comment]
It seems that an SQL statement is missing in the master branch

```
02862_sorted_distinct_sparse_fix | FAIL | 0.06
/ClickHouse/tests/queries/0_stateless/02862_sorted_distinct_sparse_fix.reference
+DROP TABLE t_sparse_distinct;
```

**Comment by marising** [comment]
When ASan and analyzer are enabled, an exception is thrown

```
cmake ../ -DSANITIZE=address -DENABLE_JEMALLOC=0
export CLICKHOUSE_USE_NEW_ANALYZER=1

            assert_cluster(
                cluster, "SELECT count() FROM cfs_test_insert FORMAT Values", "(1)", True
            )

E           helpers.client.QueryRuntimeException: Client failed! Return code: 33, stderr: Received exception from server (version 23.8.1):
E           Code: 33. DB::Exception: Received from 172.16.9.6:9000. DB::Exception: Cannot read all data in MergeTreeReaderCompact. Rows read: 0. Rows expected: 1.: (while reading from part /var/lib/clickhouse/disks/d_cfs/store/132/1320cb57-98a5-4098-807b-044dc0508e6c/20230821_0_0_0/ in table default.cfs_test_insert (1320cb57-98a5-4098-807b-044dc0508e6c) located on disk d_cfs of type cfs, from mark 0 with max_rows_to_read = 1): While executing MergeTreeInOrder. Stack trace:
E
```

File: src/IO/CFS/WriteBufferFromCFS.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#define O_DIRECT 00040000
+#endif
+
+namespace DB
+{
+/**
+ * Accepts CFS path to file and opens it.
+ * Closes file by himself (thus "owns" a file descriptor).
+ */
+class WriteBufferFromCFS final :  public WriteBufferFromFileBase
```
Comment (azat) [inline src/IO/CFS/WriteBufferFromCFS.h:23]: What are the differences with `WriteBufferFromFileDescriptor`? Can you please describe them?

Also I've noticed that this implementation does not have profile events, like `WriteBufferFromFileDescriptor` (i.e. `ProfileEvents::WriteBufferFromFileDescriptorWriteFailed`)

**Comment by marising** [comment]
AST fuzzer (debug) failed,  #54352

```
Assertion `new_head == holder.key.data' failed

2023.09.07 10:31:42.048807 [ 689 ] {} <Fatal> BaseDaemon: 9.1. inlined from /build/src/Common/HashTable/HashTableKeyHolder.h:132: keyHolderDiscardKey(DB::SerializedKeyHolder&)


inline void ALWAYS_INLINE keyHolderDiscardKey(DB::SerializedKeyHolder & holder)
{
    [[maybe_unused]] void * new_head = holder.pool.rollback(holder.key.size);
    assert(new_head == holder.key.data);  # failed
    holder.key.data = nullptr;
    holder.key.size = 0;
}

```

**Comment by danthegoodman1** [comment]
Curious why this is needed if https://github.com/ClickHouse/ClickHouse/issues/31113 seemed to work in testing? I'm not super familiar with CFS/NFS so excuse my naiveness

**Comment by marising** [comment]
> Curious why this is needed if #31113 seemed to work in testing? I'm not super familiar with CFS/NFS so excuse my naiveness

The EFS he used for this test was used as a local server disk. Different directories were used between multiple replica. It was non-shared and did not use the zero copy feature.

**Comment by danthegoodman1** [comment]
> > Curious why this is needed if #31113 seemed to work in testing? I'm not super familiar with CFS/NFS so excuse my naiveness
> 
> The EFS he used for this test was used as a local server disk. Different directories were used between multiple replica. It was non-shared and did not use the zero copy feature.

Thank you! Makes sense

**Comment by royxact** [comment]
Hi @marising ,
This issue looks very interesting, anything we can do to help get this done?

**Comment by marising** [comment]
> Hi @marising , This issue looks very interesting, anything we can do to help get this done?

Thanks for your reminding. This PR supports the CFS protocol and has the same architecture as the S3 protocol, it can expand the usage scenarios of storage and computing separation. 

In the future, it is necessary to comprehensively test and try out all the features of ClickHouse to ensure that it works normally and high-performance under storage and computation separation.

@alexey-milovidov  Do you have time to take a look at this PR and give me some suggestions？

**Comment by sunny19930321** [comment]
> This PR seems quite interesting to me as well. Looking forward to see it being merged. Maybe I mis-read, but it seems one of the checks failed?

@den-crane what can we do to make this pr merge?

**Comment by albertcht** [comment]
This feature looks really amazing and promising.
I was wondering if this PR can be merged in the near future?

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -687,6 +687,8 @@ void StorageReplicatedMergeTree::createNewZooKeeperNodes()
         futures.push_back(zookeeper->asyncTryCreateNoThrow(zookeeper_path + "/zero_copy_s3/shared", String(), zkutil::CreateMode::Persistent));
         futures.push_back(zookeeper->asyncTryCreateNoThrow(zookeeper_path + "/zero_copy_hdfs", String(), zkutil::CreateMode::Persistent));
         futures.push_back(zookeeper->asyncTryCreateNoThrow(zookeeper_path + "/zero_copy_hdfs/shared", String(), zkutil::CreateMode::Persistent));
+        futures.push_back(zookeeper->asyncTryCreateNoThrow(zookeeper_path + "/zero_copy_cfs", String(), zkutil::CreateMode::Persistent));
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:690]: A question - why do we use different names of the nodes for different types of disks?

File: src/IO/CFS/WriteBufferFromCFS.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace DB
+{
+/**
+ * Accepts CFS path to file and opens it.
+ * Closes file by himself (thus "owns" a file descriptor).
+ * The NFS protocol is compatible with most POSIX interface,
+ * but there are still differences in performance and local file systems.
+ */
+class WriteBufferFromCFS final :  public WriteBufferFromFileBase
```
Comment (alexey-milovidov) [inline src/IO/CFS/WriteBufferFromCFS.h:25]: What the difference actually is?
If it is mostly copy-paste of WriteBufferFromFile, can we unify it?

File: src/IO/CFS/ReadBufferFromCFS.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+#include <IO/ReadBufferFromFileBase.h>
+
+
+namespace DB
+{
+
+/**
+ * Accepts CFS path to file and opens it.
+ * Closes file by himself (thus "owns" a file descriptor).
+ * The NFS protocol is compatible with most POSIX interface,
```
Comment (alexey-milovidov) [inline src/IO/CFS/ReadBufferFromCFS.h:19]: .

File: src/Disks/ObjectStorages/CFS/CFSObjectStorage.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    }
+};
+
+/**
+ * Storage for persisting data in CFS and metadata on the local disk.
+ * Files are represented by file in local filesystem (clickhouse_root/disks/disk_name/path/to/file)
+ * that contains CFS object key with actual data.
+ * CFS path example : /root_dir/volume/shard/shard_number/
+ */
+class CFSObjectStorage final : public IObjectStorage
```
Comment (alexey-milovidov) [inline src/Disks/ObjectStorages/CFS/CFSObjectStorage.h:40]: Can we use LocalObjectStorage?

**Comment by alexey-milovidov** [comment]
The code can be greatly simplified (down to near zero), if we don't introduce new Read/Write Buffers, and don't introduce any new object storage types but instead reuse the LocalObjectStorage.

Let's add a functional test with replicas on the save server sharing a disk on top of the same directory on local filesystem.

**Comment by marising** [comment]
> The code can be greatly simplified (down to near zero), if we don't introduce new Read/Write Buffers, and don't introduce any new object storage types but instead reuse the LocalObjectStorage.
> 
> Let's add a functional test with replicas on the save server sharing a disk on top of the same directory on local filesystem.

There are many storage products that support the CFS , and there are some differences in compatibility and performance. 
1. Compatibility means that some Posix functions may not be supported, and alternatives need to be found.
2. Performance read and write latency from small to large: local disk < block storage(such as AWS EBS) < file system(such as AWS EFS). The difference between block storage and local disk is small, and Read/Write BufferFromFile can be used directly without any problem. The performance of the file system is much worse than that of the local disk, I remember that the `seek()`  was particularly slow in the file system, and special optimization is required for some functions with poor performance. 

Considering the scalability of the functions and performance of the CFS in the future, the code was separated with reference to the HDFS and S3 codes. After the community has fully tested and improved it, we will see whether to integrate the CFS and local file code. Do you think this plan is okay? @alexey-milovidov 

For integration testing, I used multiple replicated node to use the same config and directory to simulate a zero-copy 
remote file system, The test code is in the `test_replicated_merge_tree_cfs` directory.

**Comment by alexey-milovidov** [comment]
No need for fully separated code - let's allow the required customizations to the local filesystem type instead.

**Comment by marising** [comment]
> No need for fully separated code - let's allow the required customizations to the local filesystem type instead.

Ok, let's first test the function and performance of the local file system code to support zero copy CFS , and then see how to customize it.

**Comment by alexey-milovidov** [comment]
See https://github.com/ClickHouse/ClickHouse/pull/58357

**Comment by alexey-milovidov** [comment]
After https://github.com/ClickHouse/ClickHouse/pull/58357 it looks redundant, but we can test the performance for comparison.

**Comment by danthegoodman1** [comment]
@alexey-milovidov can we get an example of what that looks like with the new above PR features?

**Comment by alexey-milovidov** [comment]
https://www.youtube.com/live/pBF9g0wGAGs?si=u-l4ztHGI1TGmT8S&t=3116

**Comment by royxact** [comment]
Thanks @alexey-milovidov <https://github.com/alexey-milovidov> !
So for "CFS" we should use a cache disk type, wrapping a disk with
local_object_storage & local metadata?

**Comment by danthegoodman1** [comment]
> https://www.youtube.com/live/pBF9g0wGAGs?si=u-l4ztHGI1TGmT8S&t=3116

@alexey-milovidov can that code get put into docs? The linked PR doesn't have any docs and having many examples of configurations users might want is really important for adoption.

**Comment by alexey-milovidov** [comment]
@danthegoodman1, we didn't write the docs for this option, and it is fairly low-level - it is flexible, but not every combination of options makes sense, and we didn't test every combination. For example, "web" metadata on top of "local" object storage does not make sense. Let's explore these combinations together.

**Comment by danthegoodman1** [comment]
@alexey-milovidov even just listing the fact that it's an option and the available values the options can be is a fairly reasonable minimum to expect, otherwise it's really hard to explore and understand what the behavior of each option does. For example what does `plain` metadata do? when you have `web` at what level in the directory should metadata exists, and does it have to be static?

For having metadata type local like you mention above to replace this need, how does that prevent copying to replicas?

**Comment by alexey-milovidov** [comment]
Here is the demo of the `web` metadata: https://github.com/ClickHouse/ClickHouse/pull/53629#issuecomment-1937116377

**Comment by alexey-milovidov** [comment]
Here is the RFC for the `web` metadata: https://github.com/ClickHouse/ClickHouse/issues/23982

**Comment by alexey-milovidov** [comment]
Here is the performance test of the `web` metadata: https://github.com/ClickHouse/ClickBench/tree/main/clickhouse-web

**Comment by alexey-milovidov** [comment]
Here is the RFC for the `plain` metadata: https://github.com/ClickHouse/ClickHouse/pull/42232

**Comment by alexey-milovidov** [comment]
@royxact, please reopen when the performance test will be ready.

**Comment by danthegoodman1** [comment]
@alexey-milovidov i don’t intend to be irritating, as I’m just trying to make the clickhouse usage and adoption experience better because I _really_ believe in clickhouse, but you’ve just highlighted the exact problem.

You have to manage to find 4 different links, and parse through multiple PRs to understand something that could be described in <5 minutes of writing markdown docs.

The ratio of time spent writing docs to clickhouse users having to spend time digging through the GitHub and ask the maintainers questions to understand the features is obscene.

**Comment by alexey-milovidov** [comment]
@danthegoodman1, We are only starting with the feature "flexible configuration of object storages and metadata" - the docs are not written, and the feature is not ready for general usage. Please hold tight before the docs will be ready. Some enthusiasts might find my examples useful to try, but in general, it is not ready.

I recommend using only the features where the docs are already written.

**Comment by marising** [comment]
@alexey-milovidov 
1. The TPC-H query time of CFS is 10% less than that of Local. The write time of different data sizes is similar. I am not sure why the difference between the two.
2. From the perspective of architecture, CFS and Local disk look the same, but in fact, the IO is different. We will retain the support of CFS within our company.

**Pull request closed without merging**